### PR TITLE
Birmingham | 26-ITP-Jan | Tasleem Adedokun | Sprint 3 | Coursework 3 implement and rewrite

### PR DESCRIPTION
**Pull request by tasleemadedokun**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

Brief;
I have done all the coursework to the best of my ability, and I  made several commits to effect changes.

### CODE REVIEW

File: Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js [hunk 1]
```diff
@@ -11,7 +11,11 @@
 // execute the code to ensure all tests pass.
 
 function isProperFraction(numerator, denominator) {
-  // TODO: Implement this function
+       if (denominator === 0){
+        return false;
+       }
+       
+        return Math.abs(numerator) < Math.abs(denominator);
 }
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js]: Indentation is off.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js [hunk 2]
```diff
@@ -14,7 +14,30 @@ test(`should return "Acute angle" when (0 < angle < 90)`, () => {
...
+  expect(getAngleType(181)).toEqual("Reflex angle");
+  expect(getAngleType(270)).toEqual("Reflex angle");
+  expect(getAngleType(359)).toEqual("Reflex angle");
+});
+
 // Case 6: Invalid angles
+test(`should return "Invalid angle" when angle is 0 or negative`, () => {
+  expect(getAngleType(0)).toEqual("Invalid angle");
+   expect(getAngleType(360)).toEqual("Invalid angle");
+  })
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: It is good that you include both boundary cases in this test. You could also consider testing some values not on the boundary.

Shouldn't this category also include `angle >= 360`?

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js [hunk 3]
```diff
@@ -11,8 +11,25 @@ test(`Should return 11 when given an ace card`, () => {
 
 // Suggestion: Group the remaining test data into these categories:
 //   Number Cards (2-10)
+test("Should return the correct value for number cards", () => {
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js]: How about "... return the numeric value ..."?

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js [hunk 4]
```diff
@@ -11,8 +11,25 @@ test(`Should return 11 when given an ace card`, () => {
...
+  expect(getCardValue("Q♠")).toEqual(10);
+  expect(getCardValue("K♠")).toEqual(10);
+});
+
+
 //   Invalid Cards
+test("Should throw an error for invalid cards", () => {
+  expect(() => getCardValue(null)).toThrow();
+  expect(() => getCardValue(undefined)).toThrow();
+})
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js]: Why not test also strings that are not in the expected format?

**Review by cjyuan — COMMENTED** [review]
- Functions are correct

- Tests are also quite comprehensive

- Some test descriptions could be made clearer

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js [hunk 5]
```diff
@@ -8,3 +8,30 @@ const isProperFraction = require("../implement/2-is-proper-fraction");
...
+  expect(isProperFraction(-2, -4)).toEqual(true);
+})
+
+// Case 4: improper fractions (numerator >= denominator)
+test(`should return false for improper fractions`, () => {
+  expect(isProperFraction(4, 2)).toEqual(false);
+  expect(isProperFraction(4, 4)).toEqual(false);
+  expect(isProperFraction(-6, 4)).toEqual(false);
+  expect(isProperFraction(7, -2)).toEqual(false);
+});
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js]: We can use pseudo-code and notations like `abs(...)` or `| ... |` in the descriptions to more concisely describe the conditions (the "when" part).

File: Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js (hunk 1 above)
Comment (tasleemadedokun) [reply Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js]: > Indentation is off.

I have used the Indentation correctly.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js (hunk 2 above)
Comment (tasleemadedokun) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: > It is good that you include both boundary cases in this test. You could also consider testing some values not on the boundary.
> 
> Shouldn't this category also include `angle >= 360`?

I have done the changes

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js (hunk 3 above)
Comment (tasleemadedokun) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js]: > How about "... return the numeric value ..."?

perfectly implemented.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js (hunk 4 above)
Comment (tasleemadedokun) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js]: > Why not test also strings that are not in the expected format?

I have now tested for strings.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js (hunk 5 above)
Comment (tasleemadedokun) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js]: > We can use pseudo-code and notations like `abs(...)` or `| ... |` in the descriptions to more concisely describe the conditions (the "when" part).

I have done that.

Comment (tasleemadedokun) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js]: @cjyuan  Thank you very much always.  I always appreciate your reviews, and I have implemented all the changes. Thank you very much.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js [hunk 6]
```diff
@@ -11,10 +11,26 @@ test(`Should return 11 when given an ace card`, () => {
...
+test("Should return 10 for face cards (J, Q, K)", () => {
+  expect(getCardValue("J♠")).toEqual(10);
+  expect(getCardValue("Q♠")).toEqual(10);
+  expect(getCardValue("K♠")).toEqual(10);
+});
+
 //   Invalid Cards
+test("Should throw an error for invalid cards", () => {
+  expect(() => getCardValue(null)).toThrow();
+  expect(() => getCardValue(undefined)).toThrow();
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js:29]: Your code didn't actually checked if the value of card is string or not.

Your function would throw a TypeError (kind of an unexpected error) instead of the controlled "Invalid card" error. 

Including these tests in the test script may send the wrong message that the function should properly check the type of the argument and classify any value that is not a string as "Invalid card".

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js [hunk 7]
```diff
@@ -11,10 +11,26 @@ test(`Should return 11 when given an ace card`, () => {
...
+  expect(() => getCardValue("11")).toThrow();
+  expect(() => getCardValue("happy")).toThrow();
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js:31]: Could also consider testing these cases:
- Valid rank + invalid suit character
- Valid rank + no suit
- Empty string
- Valid suit + invalid rank

**Review by cjyuan — COMMENTED** [review]
Changes look good. I will mark this PR as complete first.

**Comment by illicitonion** [comment]
Closing PR because the January ITP run has finished. Feel free to re-open if you're still working on it.
